### ISSUE TITLE
Kimi-K3: qualified distribution-fidelity reference and QSRT K2 routed-expert comparison

### ISSUE THREAD

**Issue by voipmonitor** [voipmonitor/rtx6kpro#69]
## TL;DR

**Status: qualified** for paired, teacher-forced distribution-fidelity comparison against the official Kimi-K3 MXFP4 checkpoint.

- Artifact: [festr2/kimi-k3-distribution-fidelity-1024x2048-v1@402919ae](https://huggingface.co/datasets/festr2/kimi-k3-distribution-fidelity-1024x2048-v1/tree/402919ae70d61396087571b63fe9185d95491afb)
- Reproduction and candidate-comparison runbook: [models/kimi-k3/distribution-fidelity-1024x2048.md](https://github.com/local-inference-lab/rtx6kpro/blob/master/models/kimi-k3/distribution-fidelity-1024x2048.md)
- Capture implementation: [research/kimi-k3-kld-hidden-capture-20260813@e77ee061](https://github.com/local-inference-lab/vllm/tree/e77ee0612b9b7d117439920ef81bdbb162d09cd3)
- Wiki and tools publication: [local-inference-lab/rtx6kpro@442c839](https://github.com/local-inference-lab/rtx6kpro/commit/442c839c409e540b80b0714a0f5ad31d424e557c)

The artifact contains 1,024 distinct 2,048-token source contexts, 2,096,128 scored positions, 827 source clusters, BF16 pre-LM-head reference tensors, the canonical BF16 LM head, and three 64-context live-logit sentinel captures. The full payload is 164,956,945,149 bytes.

## Measurement contract

The primary metric is:

```text
KL(official MXFP4 distribution || candidate distribution)
```

Each stored reference tensor is BF16 `[2047, 7168]`, after final RMSNorm and immediately before LM-head multiplication. The comparator replays one frozen BF16 `[163840, 7168]` LM head for both operands and computes complete-vocabulary KLD, Jensen–Shannon divergence, top-1 agreement, tail statistics, depth buckets, and allocation-stratified source-cluster bootstrap intervals.

The suite has 768 analysis contexts and 256 qualification contexts. Codec parameters and acceptance thresholds must be frozen from analysis results before qualification outputs are used.

## Immutable identities

| Object | Identity |
|---|---|
| Suite manifest SHA-256 | `f3a79f7f28365d406a19a82cf210c25adf18974c4b9b607ab3754e9939f941cf` |
| Aggregate token SHA-256 | `70cd72175fcb4574123b90a83e2b71bb460b8c2938cec9573645dc8c9a46a3bd` |
| Artifact manifest SHA-256 | `2e17ac50189228ab07598e42fa3dc1a0f6b55b3653d300cb006e9c8d59c52047` |
| Official checkpoint | `moonshotai/Kimi-K3@2496450e92e425c886db095102a52a6682ca3970` |
| LM-head file SHA-256 | `c282a205aed4d19fbbe4ff907feff979c50b24723e8f552ddc141ef008c5de11` |
| Capture runtime SHA-256 | `ec02e1633b045d9e7d525b41a02f20b2e1ab08adceabb4338a678b716039a783` |
| Runtime image | `voipmonitor/vllm@sha256:974edc237f27a4eaa83a53ce4927dd176a5ad8ce4fbb8d3d689fce82348531a5` |

The official reference used TP16/DCP1, B12X MLA/MoE/linear kernels, Triton KDA prefill, BF16 activations and KV cache, and InstantTensor. Loading 1.42 TiB of checkpoint data took 167.32 seconds.

## Qualification evidence

- The 32-context live-logit compatibility suite was reproduced through offline LM-head replay with mean KLD `1.229325e-6`, top-1 agreement `0.999954`, and a vocabulary-chunk sensitivity of `1.49e-9`.
- Three official-checkpoint repeats over 64 sentinels produced mean KLD `0.0032166686`, `0.0031814546`, and `0.0031337795`; top-1 agreement ranged from `0.98326056` to `0.98348192`.
- Hidden-state replay and retained live logits produced identical metrics for all three repeat pairs.
- The complete 164.96 GB payload passed structural validation and SHA-256 verification.
- Exact normalized-content scans found zero complete-item overlap with pinned HumanEval, MMLU-test, and GPQA-Diamond revisions.

The measured official-runtime repeat variation is larger than 0.00025 KLD. A single candidate capture is therefore unsupported for claims at that resolution. Candidate differences comparable to repeat variation require repeated paired captures and source-cluster confidence intervals.

## Qualified QSRT K2 routed-expert comparison

Status: **qualified** for isolating QSRT K2 routed-expert error from non-expert
tensor storage.

The candidate combines the routed experts from
`lukealonso/Kimi-K3-QSRT-K2@3b98114115f1d41ce7963ba346c3fca19918b0bd`
with official BF16 embeddings, attention and KDA projections, normalization
weights, routers, shared experts, and LM-head tensors from
`moonshotai/Kimi-K3@2496450e92e425c886db095102a52a6682ca3970`.
The matched control uses the same QSRT K2 experts and MXFP8 non-expert tensors.

| Non-expert storage | Micro KLD | Macro-stratum KLD | Top-1 agreement |
|---|---:|---:|---:|
| Official BF16 | `0.06538554` | `0.06617214` | `93.4222%` |
| MXFP8 | `0.06702088` | `0.06780095` | `93.2957%` |

Official BF16 non-expert tensors reduce micro KLD by `0.00163534`. The paired
allocation-stratified source-cluster bootstrap 95% interval is
`[-0.00177143, -0.00151198]` across 10,000 resamples. BF16 has lower
per-context KLD in 884 of 1,024 contexts.

Immutable receipts are published under
[`results/qsrt-k2-routed-experts-official-bf16-nonexpert`](https://huggingface.co/datasets/festr2/kimi-k3-distribution-fidelity-1024x2048-v1/tree/402919ae70d61396087571b63fe9185d95491afb/results/qsrt-k2-routed-experts-official-bf16-nonexpert).
The full-vocabulary result receipt has SHA-256
`10e203e1f1ea0a9944b2bddb99243fcfbd18e16d35102155f5f32d20ba4724fe`;
the paired non-expert-storage comparison receipt has SHA-256
`5e43926ed7912dc978350899fefb2c9e225b26905768ddcf018a7887b172dbcf`.

[QSRT pull request #8](https://github.com/local-inference-lab/qsrt/pull/8)
implements an opt-in BF16 non-expert model view. It validates checkpoint
identity, tensor inventory, shapes, dtypes, serialized bits, file sizes, and
hashes. Canonical MXFP8 model views remain the default.

## Scope and limitations

The artifact measures teacher-forced checkpoint distribution fidelity at a 2,048-token context. It does not establish benchmark correctness, free-running generation quality, multimodal behavior, tool execution, long-context fidelity, speculative-decoding acceptance, or absence of cumulative routing drift.

## Review requested

Please review:

1. the environment-gated V2 runner capture hook and its global-rank-zero tensor assembly;
2. the source registry, deterministic partitioning, and 827-cluster bootstrap contract;
3. the interpretation of candidate improvements relative to measured official-runtime repeat variation.
